diffrax/_integrate.py

```diff
@@ -977,8 +1071,36 @@ def _allocate_output(subsaveat: SubSaveAt) -> SaveState:
         dense_infos=dense_infos,
         dense_save_index=dense_save_index,
         progress_meter_state=progress_meter_state,
+        tprevprev=tprevprev,
+        dense_info_for_event=dense_info_for_event,
     )
 
+    if event is not None:
+
+        def _call_event(_cond_fn):
+            return _cond_fn(
+                init_state,
+                y=y0,
+                solver=solver,
+                stepsize_controller=stepsize_controller,
+                saveat=saveat,
+                t0=t0,
+                t1=t1,
+                dt0=dt0,
+                max_steps=max_steps,
+                terms=terms,
+                args=args,
+            )
+
+        event_result = jtu.tree_map(_call_event, event.cond_fn, is_leaf=_is_cond_fn)
+        event_mask = jtu.tree_map(lambda x: False, event.cond_fn, is_leaf=_is_cond_fn)
+        init_state = eqx.tree_at(
+            lambda s: s.event_result, init_state, event_result, is_leaf=_is_none
+        )
+        init_state = eqx.tree_at(
+            lambda s: s.event_mask, init_state, event_mask, is_leaf=_is_none
+        )
```

Comment:
(a) I think if you put this before `init_state = ...` then you can just pass in this information during initialisation, without `tree_at`.
(b) I think again you can use `eval_shape` to initialise arrays of the appropriate sort without having to actually evaluate each `cond_fn`.